File: numerics/src/plotting.py
"""Figures for the article. All figures are saved in PDF and high-res PNG."""
from __future__ import annotations

from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

from .hmm_em import HMMFit
from .mixture_hmm_em import MixtureHMMFit

DPI = 300


def _save(fig, fig_dir: Path, name: str) -> None:
    fig.tight_layout()
    fig.savefig(fig_dir / f"{name}.pdf")
    fig.savefig(fig_dir / f"{name}.png", dpi=DPI)
    plt.close(fig)


def _norm_pdf(x, mu, sd):
    return np.exp(-0.5 * ((x - mu) / sd) ** 2) / (np.sqrt(2.0 * np.pi) * sd)


def plot_density(y: np.ndarray, fits: dict[int, HMMFit], fig_dir: Path) -> None:
    """Histogram of observations with fitted marginal mixture densities."""
    grid = np.linspace(y.min() - 0.8, y.max() + 0.8, 600)
    fig, ax = plt.subplots(figsize=(6.2, 3.7))
    ax.hist(y, bins=60, density=True, alpha=0.35, color="grey",
            label="observations")
    styles = {2: ("--", "C1"), 3: ("-", "C0"), 4: (":", "C2")}
    for K in sorted(fits):
        fit = fits[K]
        occ = fit.gamma.mean(axis=0)
        dens = sum(occ[j] * _norm_pdf(grid, fit.means[j], fit.sds[j])
                   for j in range(K))
        ls, c = styles.get(K, ("-", None))
        ax.plot(grid, dens, ls, color=c, lw=2, label=f"Gaussian HMM, K={K}")
    ax.set_xlabel("Observation")
    ax.set_ylabel("Density")
    ax.legend(frameon=False, fontsize=8)
    _save(fig, fig_dir, "simulation_density")


def plot_density_review(y: np.ndarray, fit2: HMMFit, fit3: HMMFit,
                        mix2: MixtureHMMFit, fig_dir: Path) -> None:
    """Histogram with Gaussian HMM and enriched-emission HMM fitted marginals."""
    grid = np.linspace(y.min() - 0.8, y.max() + 0.8, 600)
    fig, ax = plt.subplots(figsize=(6.2, 3.7))
    ax.hist(y, bins=60, density=True, alpha=0.32, color="grey",
            label="observations")
    for fit, label, ls, color in [
        (fit2, "Gaussian HMM, K=2", "--", "C1"),
        (fit3, "Gaussian HMM, K=3", "-", "C0"),
    ]:
        occ = fit.gamma.mean(axis=0)
        dens = sum(occ[j] * _norm_pdf(grid, fit.means[j], fit.sds[j])
                   for j in range(fit.K))
        ax.plot(grid, dens, ls, color=color, lw=2, label=label)
    occ = mix2.gamma.mean(axis=0)
    dens = np.zeros_like(grid)
    for k in range(mix2.K):
        for m in range(mix2.M):
            dens += occ[k] * mix2.weights[k, m] * _norm_pdf(
                grid, mix2.means[k, m], mix2.sds[k, m])
    ax.plot(grid, dens, ":", color="C2", lw=2.2,
            label="Mixture-emission HMM, K=2")
    ax.set_xlabel("Observation")
    ax.set_ylabel("Density")
    ax.legend(frameon=False, fontsize=8)
    _save(fig, fig_dir, "simulation_density")


def plot_decoding_heatmap(tab2: pd.DataFrame, tab3: pd.DataFrame,
                          fig_dir: Path) -> None:
    """Row-proportion heatmaps: true refined state vs fitted state, K=2 and K=3."""
    fig, axes = plt.subplots(1, 2, figsize=(7.2, 3.0))
    for ax, tab, K in [(axes[0], tab2, 2), (axes[1], tab3, 3)]:
        prop = tab.div(tab.sum(axis=1), axis=0)
        im = ax.imshow(prop.values, cmap="Blues", vmin=0, vmax=1, aspect="auto")
        ax.set_xticks(range(prop.shape[1]), [str(c) for c in prop.columns])
        ax.set_yticks(range(prop.shape[0]), list(prop.index))
        ax.set_xlabel(f"Fitted state (K={K})")
        if K == 2:
            ax.set_ylabel("True refined state")
        for i in range(prop.shape[0]):
            for j in range(prop.shape[1]):
                v = prop.values[i, j]
                ax.text(j, i, f"{v:.2f}", ha="center", va="center",
                        fontsize=8, color="white" if v > 0.5 else "black")
    fig.colorbar(im, ax=axes, fraction=0.035, pad=0.03)
    fig.savefig(fig_dir / "state_decoding_heatmap.pdf", bbox_inches="tight")
    fig.savefig(fig_dir / "state_decoding_heatmap.png", dpi=DPI,
                bbox_inches="tight")
    plt.close(fig)


def plot_posterior_corrective_state(y: np.ndarray, fit3: HMMFit,
                                    corrective_idx: int, fig_dir: Path) -> None:
    """Posterior probability of the corrective state as a function of Y_t."""
    fig, ax = plt.subplots(figsize=(6.2, 3.5))
    ax.scatter(y, fit3.gamma[:, corrective_idx], s=6, alpha=0.30,
               color="C0", edgecolors="none")
    # Binned mean for readability
    bins = np.quantile(y, np.linspace(0, 1, 41))
    idx = np.clip(np.digitize(y, bins) - 1, 0, 39)
    centers, means_ = [], []
    for b in range(40):
        m = idx == b
        if m.sum() >= 5:
            centers.append(y[m].mean())
            means_.append(fit3.gamma[m, corrective_idx].mean())
    ax.plot(centers, means_, color="C3", lw=2, label="binned mean")
    ax.set_xlabel(r"Observation $Y_t$")
    ax.set_ylabel("Posterior prob. of corrective state")
    ax.set_ylim(-0.03, 1.03)
    ax.legend(frameon=False, fontsize=8)
    _save(fig, fig_dir, "posterior_corrective_state")


def plot_bic_boxplot(delta_bic: np.ndarray, fig_dir: Path) -> None:
    """Boxplot of BIC(K=3) - BIC(K=2) across replications."""
    fig, ax = plt.subplots(figsize=(4.4, 3.5))
    ax.boxplot(delta_bic, vert=True, widths=0.4)
    ax.axhline(0.0, color="C3", lw=1, ls="--")
    ax.set_xticks([1], ["replications"])
    ax.set_ylabel(r"BIC$(K{=}3)$ $-$ BIC$(K{=}2)$")
    prop = float(np.mean(delta_bic < 0))
    ax.set_title(f"BIC favours K=3 in {100 * prop:.0f}% of replications",
                 fontsize=9)
    _save(fig, fig_dir, "bic_boxplot")


def plot_bic_by_T(summary: pd.DataFrame, fig_dir: Path) -> None:
    """Selection frequencies by sample size for Gaussian and enriched models."""
    fig, ax = plt.subplots(figsize=(5.2, 3.4))
    width = 0.35
    x = np.arange(len(summary))
    ax.bar(x - width / 2, summary["prop_gaussian_BIC_favours_K3"],
           width=width, color="C0", label="Gaussian: BIC favours K=3")
    ax.bar(x + width / 2, summary["prop_enriched_BIC_favours_K2_mix"],
           width=width, color="C2", label="Enriched: BIC favours K=2 mixture")
    ax.set_xticks(x, [str(int(v)) for v in summary["T"]])
    ax.set_ylim(0, 1.05)
    ax.set_xlabel("T")
    ax.set_ylabel("Selection frequency")
    ax.legend(frameon=False, fontsize=8)
    _save(fig, fig_dir, "bic_by_T")


def plot_residual_acf(acf2: pd.DataFrame, acf3: pd.DataFrame, fig_dir: Path) -> None:
    """ACF of one-step standardized prediction residuals."""
    fig, ax = plt.subplots(figsize=(5.4, 3.4))
    ax.axhline(0.0, color="black", lw=0.8)
    ax.plot(acf2["lag"], acf2["acf"], marker="o", ms=3, lw=1.2,
            color="C1", label="Gaussian HMM, K=2")
    ax.plot(acf3["lag"], acf3["acf"], marker="o", ms=3, lw=1.2,
            color="C0", label="Gaussian HMM, K=3")
    ax.set_xlabel("Lag")
    ax.set_ylabel("Residual ACF")
    ax.legend(frameon=False, fontsize=8)
    _save(fig, fig_dir, "residual_acf")

File: numerics/R/plotting.R
# ggplot2 figures for the numerical study.

save_ggplot <- function(plot, directory, name, width, height) {
  dir.create(directory, recursive = TRUE, showWarnings = FALSE)
  ggplot2::ggsave(
    filename = file.path(directory, paste0(name, ".pdf")),
    plot = plot,
    width = width,
    height = height,
    units = "in",
    device = grDevices::cairo_pdf
  )
  ggplot2::ggsave(
    filename = file.path(directory, paste0(name, ".png")),
    plot = plot,
    width = width,
    height = height,
    units = "in",
    dpi = 300
  )
  invisible(plot)
}

study_theme <- function(base_size = 9) {
  ggplot2::theme_minimal(base_size = base_size) +
    ggplot2::theme(
      panel.grid.minor = ggplot2::element_blank(),
      legend.position = "bottom",
      legend.title = ggplot2::element_blank(),
      plot.title.position = "plot"
    )
}

occupation_weighted_density <- function(grid, fit) {
  occupation <- colMeans(fit$posterior)
  if (inherits(fit, "mixture_hmm_fit")) {
    output <- numeric(length(grid))
    for (state in seq_len(fit$K)) {
      for (component in seq_len(fit$component_counts[[state]])) {
        output <- output + occupation[[state]] * fit$weights[state, component] *
          stats::dnorm(
            grid,
            mean = fit$means[state, component],
            sd = fit$sds[state, component]
          )
      }
    }
    return(output)
  }
  Reduce(
    `+`,
    lapply(seq_len(fit$K), function(state) {
      occupation[[state]] * stats::dnorm(
        grid, mean = fit$means[[state]], sd = fit$sds[[state]]
      )
    })
  )
}

plot_simulation_density <- function(y, fit2, fit3, mixture2, figure_directory) {
  grid <- seq(min(y) - 0.8, max(y) + 0.8, length.out = 600)
  density_data <- rbind(
    data.frame(
      observation = grid,
      density = occupation_weighted_density(grid, fit2),
      model = "Gaussian HMM, K=2"
    ),
    data.frame(
      observation = grid,
      density = occupation_weighted_density(grid, fit3),
      model = "Gaussian HMM, K=3"
    ),
    data.frame(
      observation = grid,
      density = occupation_weighted_density(grid, mixture2),
      model = "Mixture-emission HMM, K=2"
    )
  )
  density_data$model <- factor(
    density_data$model,
    levels = c(
      "Gaussian HMM, K=2",
      "Gaussian HMM, K=3",
      "Mixture-emission HMM, K=2"
    )
  )
  plot <- ggplot2::ggplot(data.frame(observation = y), ggplot2::aes(observation)) +
    ggplot2::geom_histogram(
      ggplot2::aes(y = ggplot2::after_stat(density)),
      bins = 60,
      fill = "grey70",
      color = "white",
      linewidth = 0.2
    ) +
    ggplot2::geom_line(
      data = density_data,
      ggplot2::aes(y = density, color = model, linetype = model),
      linewidth = 0.8
    ) +
    ggplot2::scale_color_manual(values = c("#C76E2E", "#2F6B9A", "#2D8A66")) +
    ggplot2::scale_linetype_manual(values = c("dashed", "solid", "dotted")) +
    ggplot2::labs(x = "Observation", y = "Density") +
    study_theme()
  save_ggplot(plot, figure_directory, "simulation_density", 6.2, 3.7)
}

plot_decoding_heatmap <- function(table2, table3, figure_directory) {
  make_data <- function(tab, K) {
    proportions <- prop.table(tab, margin = 1)
    output <- as.data.frame(proportions, responseName = "proportion")
    names(output)[1:2] <- c("true_refined", "fitted_state")
    output$K <- paste0("Gaussian HMM, K=", K)
    output
  }
  data <- rbind(make_data(table2, 2), make_data(table3, 3))
  plot <- ggplot2::ggplot(
    data,
    ggplot2::aes(fitted_state, true_refined, fill = proportion)
  ) +
    ggplot2::geom_tile(color = "white", linewidth = 0.5) +
    ggplot2::geom_text(
      ggplot2::aes(label = sprintf("%.2f", proportion)),
      size = 2.8
    ) +
    ggplot2::facet_wrap(~K, scales = "free_x") +
    ggplot2::scale_fill_gradient(low = "white", high = "#2F6B9A", limits = c(0, 1)) +
    ggplot2::labs(
      x = "Fitted state",
      y = "True refined state",
      fill = "Row proportion"
    ) +
    study_theme() +
    ggplot2::theme(panel.grid = ggplot2::element_blank())
  save_ggplot(plot, figure_directory, "state_decoding_heatmap", 7.2, 3.0)
}

plot_corrective_posterior <- function(
    y, fit3, corrective_state, figure_directory) {
  breaks <- unique(as.numeric(stats::quantile(
    y, seq(0, 1, length.out = 41), names = FALSE
  )))
  bin <- cut(y, breaks = breaks, include.lowest = TRUE, labels = FALSE)
  point_data <- data.frame(
    observation = y,
    posterior = fit3$posterior[, corrective_state],
    bin = bin
  )
  binned <- dplyr::summarise(
    dplyr::group_by(point_data, bin),
    observation = mean(observation),
    posterior = mean(posterior),
    n = dplyr::n(),
    .groups = "drop"
  )
  binned <- dplyr::filter(binned, n >= 5)
  plot <- ggplot2::ggplot(
    point_data,
    ggplot2::aes(observation, posterior)
  ) +
    ggplot2::geom_point(color = "#2F6B9A", alpha = 0.30, size = 0.8) +
    ggplot2::geom_line(
      data = binned,
      color = "#C44E52",
      linewidth = 0.8
    ) +
    ggplot2::coord_cartesian(ylim = c(0, 1)) +
    ggplot2::labs(
      x = expression("Observation " * Y[t]),
      y = "Posterior probability of corrective state"
    ) +
    study_theme()
  save_ggplot(plot, figure_directory, "posterior_corrective_state", 6.2, 3.5)
}

plot_bic_boxplot <- function(delta_bic, figure_directory) {
  frequency <- mean(delta_bic < 0)
  plot <- ggplot2::ggplot(
    data.frame(group = "replications", delta_bic = delta_bic),
    ggplot2::aes(group, delta_bic)
  ) +
    ggplot2::geom_hline(yintercept = 0, color = "#C44E52", linetype = "dashed") +
    ggplot2::geom_boxplot(width = 0.4, fill = "#D8E6F2", color = "#2F6B9A") +
    ggplot2::labs(
      x = NULL,
      y = expression(BIC(K == 3) - BIC(K == 2)),
      title = sprintf("BIC favours K=3 in %.1f%% of replications", 100 * frequency)
    ) +
    study_theme()
  save_ggplot(plot, figure_directory, "bic_boxplot", 4.4, 3.5)
}

plot_bic_by_sample_size <- function(summary, figure_directory) {
  data <- rbind(
    data.frame(
      T = summary$T,
      selection_frequency = summary$prop_gaussian_BIC_favours_K3,
      comparison = "Gaussian: BIC favours K=3"
    ),
    data.frame(
      T = summary$T,
      selection_frequency = summary$prop_enriched_BIC_favours_K2_mix,
      comparison = "Enriched: BIC favours K=2 mixture"
    )
  )
  plot <- ggplot2::ggplot(
    data,
    ggplot2::aes(factor(T), selection_frequency, fill = comparison)
  ) +
    ggplot2::geom_col(position = ggplot2::position_dodge(width = 0.75), width = 0.68) +
    ggplot2::scale_fill_manual(values = c("#2D8A66", "#2F6B9A")) +
    ggplot2::coord_cartesian(ylim = c(0, 1)) +
    ggplot2::labs(x = "T", y = "Selection frequency") +
    study_theme()
  save_ggplot(plot, figure_directory, "bic_by_T", 5.2, 3.4)
}

plot_residual_acf <- function(acf2, acf3, figure_directory) {
  data <- rbind(
    transform(acf2, model = "Gaussian HMM, K=2"),
    transform(acf3, model = "Gaussian HMM, K=3")
  )
  plot <- ggplot2::ggplot(data, ggplot2::aes(lag, acf, color = model)) +
    ggplot2::geom_hline(yintercept = 0, color = "grey30", linewidth = 0.4) +
    ggplot2::geom_line(linewidth = 0.6) +
    ggplot2::geom_point(size = 1.4) +
    ggplot2::scale_color_manual(values = c("#C76E2E", "#2F6B9A")) +
    ggplot2::labs(x = "Lag", y = "Residual ACF") +
    study_theme()
  save_ggplot(plot, figure_directory, "residual_acf", 5.4, 3.4)
}
